### PR TITLE
Migrate InitBuild to provider API

### PR DESCRIPTION
**Pull request by ov7a**
### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by ov7a** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/internal/provider/ProviderApiDeprecationLogger.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.internal.deprecation.DeprecationLogger;
+
+public class ProviderApiDeprecationLogger {
```
Comment (tresat) [inline subprojects/core-api/src/main/java/org/gradle/api/internal/provider/ProviderApiDeprecationLogger.java:21]: 💅 Missing javadoc on top-level type.

File: platforms/jvm/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoTaskExtension.java [hunk 2]
```diff
@@ -125,177 +126,123 @@ public File getDestinationFile() {
...
+    @ReplacesEagerProperty(originalType = boolean.class)
+    public abstract Property<Boolean> getIncludeNoLocationClasses();
 
-    public void setIncludeNoLocationClasses(boolean includeNoLocationClasses) {
-        this.includeNoLocationClasses = includeNoLocationClasses;
+    @Internal
+    @Deprecated
+    @SuppressWarnings("InlineMeSuggester")
+    public Property<Boolean> getIsIncludeNoLocationClasses() {
+        // TODO: add deprecation message
```
Comment (tresat) [inline platforms/jvm/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoTaskExtension.java:172]: 🤔 TODO on several props in this class - are these meant to be added now?

File: platforms/jvm/language-jvm/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java [hunk 3]
```diff
@@ -114,124 +85,125 @@ public CompileOptions(ObjectFactory objectFactory) {
...
-    @ToBeReplacedByLazyProperty
-    public boolean isFailOnError() {
-        return failOnError;
-    }
+    @ReplacesEagerProperty(originalType = boolean.class)
+    public abstract Property<Boolean> getFailOnError();
 
     /**
-     * Sets whether to fail the build when compilation fails. Defaults to {@code true}.
+     * TODO: Add deprecation warning
```
Comment (tresat) [inline platforms/jvm/language-jvm/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java:106]: 🤔 TODOs about missing deprecation warnings in this type

File: platforms/software/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java [hunk 4]
```diff
@@ -90,66 +96,81 @@ public enum PathBase {
...
+        getJarFile().convention(getProject().getObjects().fileProperty()).fileValue(getFileOperations().file(WrapperDefaults.JAR_FILE_PATH));
+        getDistributionPath().convention(DEFAULT_DISTRIBUTION_PARENT_NAME);
+        getDistributionBase().convention(WrapperDefaults.DISTRIBUTION_BASE);
+        getDistributionType().convention(WrapperDefaults.DISTRIBUTION_TYPE);
+        getArchivePath().convention(WrapperDefaults.ARCHIVE_PATH);
+        getArchiveBase().convention(WrapperDefaults.ARCHIVE_BASE);
+        getGradleVersion().convention(GradleVersion.current().getVersion());
+        getDistributionUrl().convention(getGradleVersion().zip(getDistributionType(), this::getDistributionUrl));
+        this.batchScript = getObjectFactory().fileProperty().fileProvider(getScriptFile().getLocationOnly().map(file -> WrapperGenerator.getBatchScript(file.getAsFile())));
+        this.propertiesFile = getObjectFactory().fileProperty().fileProvider(getJarFile().getLocationOnly().map(file -> WrapperGenerator.getPropertiesFile(file.getAsFile())));
```
Comment (tresat) [inline platforms/software/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java:117]: 💭 Could every convention use the same `getObjectFactory()` vs. `getProject().getObjects()` here?

**Review by tresat — APPROVED** [review]

**Comment by ov7a** [comment]
Wow, the rebase of the parent branch messed up the history here. I've rebased it to sync with the parent branch

**Comment by cobexer** [comment]
> Wow, the rebase of the parent branch messed up the history here. I've rebased it to sync with the parent branch

That is to be expected for all in-flight PRs after rebases.

**Comment by ov7a** [comment]
@bot-gradle test this

**Comment by ov7a** [comment]
@bot-gradle test this

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 5]
```diff
@@ -485,33 +458,33 @@ private GradleException createNotSupportedTestFrameWorkException(BuildInitialize
 
     @VisibleForTesting
     String getEffectiveProjectName(UserQuestions userQuestions, BuildInitializer initializer) {
-        String projectName = this.projectName;
         if (initializer.supportsProjectName()) {
-            if (isNullOrEmpty(projectName)) {
-                return userQuestions.askQuestion("Project name", getProjectName());
+            if (getProjectName().isPresent()) {
```
Comment (asodja) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:462]: 🤔 I wonder if you technically can have project name present and it's an empty string?

Could that be a problem? Do we need to check:
```if (isNullOrEmpty(getProjectName().getOrNull()) {}```
?

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 6]
```diff
@@ -485,33 +458,33 @@ private GradleException createNotSupportedTestFrameWorkException(BuildInitialize
...
-        String packageName = this.packageName;
         if (initializer.supportsPackage()) {
-            if (packageName == null) {
-                return getProviderFactory().gradleProperty(SOURCE_PACKAGE_PROPERTY).getOrElse(SOURCE_PACKAGE_DEFAULT);
-            }
-        } else if (!isNullOrEmpty(packageName)) {
+            return getPackageName().getOrElse(
+                getProviderFactory().gradleProperty(SOURCE_PACKAGE_PROPERTY).getOrElse(SOURCE_PACKAGE_DEFAULT)
+            );
+        } else if (getPackageName().isPresent()) {
```
Comment (asodja) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:479]: 🤔 Same here for package name, can it be present but it is an empty string?

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 7]
```diff
@@ -485,33 +458,33 @@ private GradleException createNotSupportedTestFrameWorkException(BuildInitialize
...
             throw new GradleException("Package name is not supported for '" + initializer.getId() + "' build type.");
         }
-        return packageName;
+        return "";
     }
 
     private BuildInitializer getBuildInitializer(UserQuestions userQuestions, ProjectLayoutSetupRegistry projectLayoutRegistry) {
-        if (!isNullOrEmpty(type)) {
-            return projectLayoutRegistry.get(type);
+        if (getType().isPresent()) {
```
Comment (asodja) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:486]: 🤔 Same here for type, can it be present but it is an empty string?

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 8]
```diff
@@ -198,48 +182,38 @@ public Property<Boolean> getUseIncubating() {
...
      */
     @Input
-    @ToBeReplacedByLazyProperty
-    public String getProjectName() {
-        return projectName == null ? projectDir.getAsFile().getName() : projectName;
-    }
+    @Optional
+    @Option(option = "project-name", description = "Set the project name.")
+    @ReplacesEagerProperty
+    public abstract Property<String> getProjectName();
```
Comment (asodja) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:196]: ❓ Should we set convention for getProjectName():
```
getProjectName().convention(projectDir.getAsFile().getName())
```
so it's the same as original method?

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 9]
```diff
@@ -198,48 +182,38 @@ public Property<Boolean> getUseIncubating() {
...
      */
     @Input
-    @ToBeReplacedByLazyProperty
-    public String getPackageName() {
-        return packageName == null ? "" : packageName;
-    }
+    @Optional
+    @ReplacesEagerProperty
+    @Option(option = "package", description = "Set the package for source files.")
+    public abstract Property<String> getPackageName();
```
Comment (asodja) [inline platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:207]: ❓ Should this be by convention empty?

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java (hunk 8 above)
Comment (ov7a) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:196]: It's a bit complicated. 
Initially, I did a convention, but the current logic relies on the fact that the value was set or not. With current provider API we can't say if the value was changed or not (only if the value is present, which is always true with the conventional value)

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java (hunk 9 above)
Comment (ov7a) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:207]: It's the same as for the projectName

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java (hunk 8 above)
Comment (asodja) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:196]: Oh I see, and projectDir name becomes a default when asking a question... Got it

**Review by asodja — APPROVED** [review]
LGTM

File: platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java (hunk 5 above)
Comment (ov7a) [reply platforms/software/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:462]: I think it is possible to set it to "", but in that case, this means the user specifically specified that value, and there is no reason to go into interactive mode.

**Comment by ov7a** [comment]
@bot-gradle merge

**Pull request merged**